designs/karpenter-integration-testing.md

```diff
@@ -0,0 +1,271 @@
+    * Utilization
+        * should provision one pod per node
+    * Validation 
+        * NodePool
+            * should error when a restricted label is used in labels (karpenter.sh/nodepool)
+            * should error when a restricted label is used in labels (kubernetes.io/custom-label)
+            * should allow a restricted label exception to be used in labels (node-restriction.kubernetes.io/custom-label)
+            * should allow a restricted label exception to be used in labels ([*].node-restriction.kubernetes.io/custom-label)
+            * should error when a requirement references a restricted label (karpenter.sh/nodepool)
+            * should error when a requirement uses In but has no values
+            * should error when a requirement uses an unknown operator
+            * should error when Gt is used with multiple integer values
+            * should error when Lt is used with multiple integer values
+            * should error when consolidateAfter is negative
+            * should succeed when ConsolidationPolicy=WhenEmptyOrUnderutilized is used with consolidateAfter
+            * should error when minValues for a requirement key is negative
+            * should error when minValues for a requirement key is zero
+            * should error when minValues for a requirement key is more than 50
+            * should error when minValues for a requirement key is greater than the values specified within In operator
+
+##### Nodeclaim
+
+* High Level Description: Validate that action taken by the nodeclaim controllers. This validates standalone claims and the garbage collection mechanisms  
+* Source: https://github.com/aws/karpenter-provider-aws/tree/main/test/suites/nodeclaim
+* Test Cases: 
+    * Standalone NodeClaim 
+        * should create a standard NodeClaim with the c instance family 
+        * should create a standard NodeClaim based on resource requests
+        * should create a NodeClaim propagating all NodeClaim spec details
+        * should remove the cloudProvider NodeClaim when the cluster NodeClaim is deleted
+        * should delete a NodeClaim from the node termination finalizer 
+        * should create a NodeClaim with custom labels passed through the userData
+        * should delete a NodeClaim after the registration timeout when the node doesn’t register
+        * should delete a NodeClaim if it references a NodeClass that doesn’t exist
+        * should delete a NodeClaim if it references a NodeClass that isn’t Ready 
+    * Garbage Collection 
+        * should succeed to garbage collect an Instance that was launched by a NodeClaim but has no Instance mapping 
+        * should succeed to garbage collect an Instance that was deleted without cluster’s knowledge 
+
+##### Scale
+
+* High Level Description: Scale test to see any regression with Karpenter performance. These are high level tests that only really validate the provisioning and speed of getting pods to a ready state.  
```

Comment:
Does it make sense to port over these scale tests? If these tests are to validate the speed of getting pods to a ready state, wouldn't that be a cloud-provider specific problem since I'm assuming different infrastructure have different performance benchmarks?

Or is the expectation that all cloud-providers should be able to provision nodes (I think I see `expectedNodes=500` for one of the tests) in a similar/reasonable amount of time?

Reply:
We are planning on doing some benchmarking work on our scheduler, so I'm not sure if it makes sense to move these tests from the AWS provider. @DerekFrank is planning on doing that work